File: algorithm/滑动平均算法.cpp
#include <vector>
#include <iostream>



std::vector<float> moving_average(const std::vector<float>& data, int window_size) {
    std::vector<float> smoothed_data(data.size());

    for (int i = 0; i < data.size(); i++) {
        float sum = 0;
        int count = 0;

        for (int j = i - window_size / 2; j <= i + window_size / 2; j++) {
            if (j >= 0 && j < data.size()) {
                sum += data[j];
                count++;
            }
        }

        smoothed_data[i] = sum / count;
    }

    return smoothed_data;
}


int main(int argc, char const *argv[])
{
    data = ...;
    int window_size = 9;
    std::vector<float> smoothed_data = moving_average(data, window_size);
    
    for (size_t i = 0; i < smoothed_data.size(); i++)
    {
        std::cout << smoothed_data.at(i) << std::endl;
    }
    
    return 0;
}

File: algorithm/滑动平均算法.py



def sliding_window_average(data, window_size):
    """
    Computes the sliding window average of a list of data points.

    Args:
        data (list): A list of data points.
        window_size (int): The size of the sliding window.

    Returns:
        list: A list of the sliding window averages.
    """
    if window_size > len(data):
        raise ValueError("Window size cannot be larger than data size.")
    window_sum = sum(data[:window_size])
    window_avg = window_sum / window_size
    averages = [window_avg]
    for i in range(window_size, len(data)):
        window_sum += data[i] - data[i-window_size]
        window_avg = window_sum / window_size
        averages.append(window_avg)
    return averages


if __name__ == '__main__':
    data = [0.1, 0.15, 0.05, 0.08, 0.07, 0.14, 0.12]
    window_size = 4
    data_res = sliding_window_average(data, window_size)

    for i in data_res:
        print(i)
